### PR TITLE
James 2436 Handle mailboxes in archives

### PR DESCRIPTION
**Pull request by rouazana**
Including uidvalidity and mailboxid metadata
Todo: store internalDate in messages metadata

Also some design notes: with Antoine we decided to not store folder hierarchy as a real hierarchy because it creates more problem than it solves. For the same reason, messages are not stored in their respective mailbox because if one day we want to store several mailboxIds for one message we won't be able to store messages in their mailboxes.
With this design we also do not need to care about checking that parent mailboxes are well created, nor that every mailboxes are created for every mail.

### CODE REVIEW

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/MailboxIdExtraField.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+package org.apache.james.mailbox.backup;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.apache.commons.compress.archivers.zip.ZipExtraField;
+import org.apache.commons.compress.archivers.zip.ZipShort;
+
+public class MailboxIdExtraField implements ZipExtraField {
```
Comment (chibenwa) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/MailboxIdExtraField.java]: Can't we also have a String extra-field?

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/LongExtraField.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    public Optional<Long> getValue() {
+        return value;
+    }
+
+    @Override
+    public final boolean equals(Object o) {
+        if (o instanceof LongExtraField) {
+            LongExtraField that = (LongExtraField) o;
+
+            return Objects.equals(this.value, that.value);
```
Comment (chibenwa) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/LongExtraField.java]: We should add the ZipShort ID overwize, SizeExtraField can be equal to UidExtraField!

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipperTest.java [hunk 3]
```diff
@@ -90,27 +91,91 @@ void archiveShouldWriteTwoMessagesWhenTwo() throws Exception {
...
+    @Test
+    void archiveShouldWriteMailboxMetadataWhenPresent() throws Exception {
+        testee.archive(ImmutableList.of(MAILBOX_1), Stream.of(), output);
+
+        try (ZipFile zipFile = new ZipFile(toSeekableByteChannel(output))) {
+            assertThatZip(zipFile)
+                .containsOnlyEntriesMatching(
+                    hasName(MAILBOX_1.getName() + "/")
+                        .containsExtraFields(
+                            new MailboxIdExtraField(MAILBOX_1.getMailboxId().serialize()),
```
Comment (chibenwa) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipperTest.java]: MailboxIdExtraField should have  MailboxId constructor, no?

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipperTest.java [hunk 4]
```diff
@@ -90,27 +91,91 @@ void archiveShouldWriteTwoMessagesWhenTwo() throws Exception {
...
-        try (ZipFile zipFile = new ZipFile(destination)) {
+        try (ZipFile zipFile = new ZipFile(toSeekableByteChannel(output))) {
             assertThatZip(zipFile)
                 .containsOnlyEntriesMatching(
-                    hasName(MESSAGE_ID_2.serialize())
-                        .hasStringContent(MESSAGE_CONTENT_2));
+                    hasName(MESSAGE_ID_1.serialize())
+                        .containsExtraFields(new SizeExtraField(SIZE_1))
+                        .containsExtraFields(new UidExtraField(MESSAGE_UID_1_VALUE))
+                        .containsExtraFields(new MessageIdExtraField(MESSAGE_ID_1.serialize()))
```
Comment (chibenwa) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipperTest.java]: Should have a MessageId constructor?

**Comment by trantienduchn** [comment]
I could not push commits to @rouazana remote, then I pushed to my remote `trantienduchn`:`JAMES-2436`

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/MailboxIdExtraField.java [hunk 5]
```diff
@@ -19,87 +19,33 @@
...
-        return mailboxId
-            .map(value -> value.getBytes(StandardCharsets.UTF_8).length)
-            .map(ZipShort::new)
-            .orElseThrow(() -> new RuntimeException("Value must by initialized"));
-    }
-
-    @Override
-    public ZipShort getCentralDirectoryLength() {
-        return getLocalFileDataLength();
+        super();
```
Comment (chibenwa) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/MailboxIdExtraField.java]: All your constructors should call the optional one...

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/MessageIdExtraField.java [hunk 6]
```diff
@@ -19,87 +19,33 @@
...
-        return messageId
-            .map(value -> value.getBytes(StandardCharsets.UTF_8).length)
-            .map(ZipShort::new)
-            .orElseThrow(() -> new RuntimeException("Value must by initialized"));
-    }
-
-    @Override
-    public ZipShort getCentralDirectoryLength() {
-        return getLocalFileDataLength();
+        super();
```
Comment (chibenwa) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/MessageIdExtraField.java]: Idem

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/LongExtraField.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+            return Objects.equals(this.value, that.value)
+                && Objects.equals(this.getHeaderId(), that.getHeaderId());
+        }
+        return false;
+    }
+
+    @Override
+    public final int hashCode() {
+        return Objects.hash(value);
```
Comment (chibenwa) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/LongExtraField.java]: HashCode don't depend on headerId?

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/StringExtraField.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    @Override
+    public final boolean equals(Object o) {
+        if (this.getClass() == o.getClass()) {
+            StringExtraField that = (StringExtraField) o;
+
+            return Objects.equals(this.getValue(), that.getValue())
+                && Objects.equals(this.getHeaderId(), that.getHeaderId());
+        }
+        return false;
+    }
```
Comment (chibenwa) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/StringExtraField.java]: No hashCode ?

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java [hunk 9]
```diff
@@ -19,29 +19,29 @@
...
         super();
     }
 
-    public InternalDateExtraField(long time) {
-        super(time);
-    }
-
     public InternalDateExtraField(Date date) {
-        super(date.getTime());
+        super(Optional.of(date.toInstant().toString()));
```
Comment (chibenwa) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: Use the Optional constructor...

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java [hunk 10]
```diff
@@ -19,29 +19,29 @@
...
     }
 
     public InternalDateExtraField(Optional<Date> date) {
-        super(date.map(Date::getTime));
+        super(date
+            .map(Date::toInstant)
```
Comment (chibenwa) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: By using $Instant$ we loose the information about the timezone.

Furstermore, we should avoid "toString" calls.

I propose to rely on : 

```
DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(zonedDateTime)
ZonedDateTime.parse(dateAsString, DateTimeFormatter.ISO_OFFSET_DATE_TIME);
And how the date looks like: "2011-12-03T10:15:30+01:00"
```

Comment (trantienduchn) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: do you want to use `Optional<ZonedDateTime>` instead of `Optional<Date>` as an constructor parameter? Since, `message.getInternalDate()` returns a `Date`, then we do not have timezone information. Suppose we use the machine time zone. I pushed updated into `trantienduchn`:`JAMES-2436`

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/Backup.java [hunk 11]
```diff
@@ -21,10 +21,17 @@
...
 
 public interface Backup {
 
-    void archive(List<MailboxMessage> messages, OutputStream destination) throws IOException;
+    /**
+     * @param mailboxes list of mailboxes to be stored in the archive
+     * @param messages a stream of MailboxMessages that will be consumed
+     * @param destination an OutputStream in which the zip will be written
+     */
+    void archive(List<Mailbox> mailboxes, Stream<MailboxMessage> messages, OutputStream destination) throws IOException;
```
[the remark this answers is not in the document]
Comment (rouazana) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/Backup.java]: yes but in this case we did not find a better way to share this big chunk of data. Do you have something better?
BTW that's why the documentation explicitly states that the stream will be consumed

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+package org.apache.james.mailbox.backup;
+
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.util.Date;
+import java.util.Optional;
+
+import org.apache.commons.compress.archivers.zip.ZipShort;
+
+public class InternalDateExtraField extends StringExtraField {
```
[the remark this answers is not in the document]
Comment (aduprat) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: Maybe next time...
I think that bringing a template solution in order to get rid of a few lines is maybe too much.

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/MailboxIdExtraField.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+ ****************************************************************/
+
+package org.apache.james.mailbox.backup;
+
+import java.util.Optional;
+
+import org.apache.commons.compress.archivers.zip.ZipShort;
+import org.apache.james.mailbox.model.MailboxId;
+
+public class MailboxIdExtraField extends StringExtraField {
```
[the remark this answers is not in the document]
Comment (aduprat) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/MailboxIdExtraField.java]: idem

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/Backup.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (mbaechler) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/Backup.java]: in fact the Stream abstraction is not ideal because it means that the stream magically calls a backend to load more data and the archive method has to handle failures of the backend it knows nothing about.
A much better abstraction would be to use a RxJava-like (reactive stream) : the backend could then push data in the reactive stream and have backup handling entries without sacrifying extensibility (you can have several consumer to the reactive stream).
It's probably not for today but probably something to think about for long running tasks.

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/Zipper.java [hunk 14]
```diff
@@ -37,15 +39,25 @@ public Zipper() {
     }
 
     @Override
-    public void archive(Stream<MailboxMessage> messages, OutputStream destination) throws IOException {
+    public void archive(List<Mailbox> mailboxes, Stream<MailboxMessage> messages, OutputStream destination) throws IOException {
         try (ZipArchiveOutputStream archiveOutputStream = new ZipArchiveOutputStream(destination)) {
+            for (Mailbox mailbox: mailboxes) {
```
Comment (mbaechler) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/Zipper.java]: I'd like this loop to be inside a storeMailbox method and the following one inside a storeMessages method

**Comment by chibenwa** [comment]
```
[a53a33c3931576aa074b50c5bedb546dd1e675d2] Results :
[a53a33c3931576aa074b50c5bedb546dd1e675d2] 
[a53a33c3931576aa074b50c5bedb546dd1e675d2] Failed tests: 
[a53a33c3931576aa074b50c5bedb546dd1e675d2]   InternalDateExtraFieldTest$GetCentralDirectoryData.getCentralDirectoryDataShouldReturnDateByteArrayWhenPassDate:172 expected:<...50, 45, 49, 53, 84, [50, 50, 58, 53, 52, 58, 48, 50, 43, 48, 55, 58, 48, 48]]> but was:<...50, 45, 49, 53, 84, [49, 53, 58, 53, 52, 58, 48, 50, 90]]>
[a53a33c3931576aa074b50c5bedb546dd1e675d2]   InternalDateExtraFieldTest$GetLocalFileDataData.getLocalFileDataShouldReturnDateByteArrayWhenPassDate:139 expected:<...50, 45, 49, 53, 84, [50, 50, 58, 53, 52, 58, 48, 50, 43, 48, 55, 58, 48, 48]]> but was:<...50, 45, 49, 53, 84, [49, 53, 58, 53, 52, 58, 48, 50, 90]]>
```

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Strange failure without output in 

```
[9a0429888431f21bc8de7d070fd8cfca931613fb] [INFO] Apache James :: Server :: JMAP :: Integration testing Common FAILURE [  02:00 h]
```

test this please

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/InternalDateExtraFieldTest.java [hunk 15]
```diff
@@ -112,152 +124,212 @@ void getCentralDirectoryLengthShouldReturnIntegerSize() {
...
 
-            testee.parseFromCentralDirectoryData(DATE_STRING_1_BYTE_ARRAY, 0, 20);
+            Date expectedTimeZonedDate = Date.from(ZonedDateTime
+                .ofInstant(DEFAULT_DATE.toInstant(), ZoneId.systemDefault())
+                .toInstant());
 
-            assertThat(testee.getDateValue())
-                .contains(Date.from(DATE_1.toInstant()));
+            assertThat(testee.getLocalDateValue())
+                .contains(expectedTimeZonedDate);
```
[the remark this answers is not in the document]
Comment (trantienduchn) [reply mailbox/backup/src/test/java/org/apache/james/mailbox/backup/InternalDateExtraFieldTest.java]: my bad,

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    public static final ZipShort ID = new ZipShort(0x6F61); // "ao" in little-endian
+
+    public InternalDateExtraField() {
+        super();
+    }
+
+    public InternalDateExtraField(Optional<Date> date) {
+        super(date
+            .map(Date::toInstant)
+            .map(instant -> ZonedDateTime.ofInstant(instant, ZoneId.systemDefault()))
```
Comment (rouazana) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: I don't get why you need to convert this Instant again and again. Can you explain me why?
I would have expected
```
date.map(Date::getTime)
```
to be sufficient

Comment (rouazana) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: (instead of these 6 lines)

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+            .map(ZonedDateTime::toInstant)
+            .map(Instant::toEpochMilli));
+    }
+
+    public InternalDateExtraField(Date date) {
+        this(Optional.of(date));
+    }
+
+    public InternalDateExtraField(long timestamp) {
+        this(Optional.of(new Date(timestamp)));
```
Comment (rouazana) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: BTW that what you are doing here

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+        return getValue().map(Date::new);
+    }
+
+    public Optional<Date> getLocalDateValue() {
+        return getValue()
+            .map(Instant::ofEpochMilli)
+            .map(instant -> ZonedDateTime.ofInstant(instant, ZoneId.of("UTC")))
+            .map(utcTime -> utcTime.withZoneSameInstant(ZoneId.systemDefault()))
+            .map(ZonedDateTime::toInstant)
+            .map(Date::from);
```
Comment (rouazana) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: but not here...

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/Zipper.java [hunk 19]
```diff
@@ -22,32 +22,69 @@
...
+    private void storeMailboxes(List<Mailbox> mailboxes, ZipArchiveOutputStream archiveOutputStream) throws IOException {
+        for (Mailbox mailbox: mailboxes) {
+            storeInArchive(mailbox, archiveOutputStream);
+        }
+    }
+
+    private void storeMessages(Stream<MailboxMessage> messages, ZipArchiveOutputStream archiveOutputStream) {
+        messages.forEach(Throwing.<MailboxMessage>consumer(message -> {
+            storeInArchive(message, archiveOutputStream);
+        }).sneakyThrow());
```
Comment (rouazana) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/Zipper.java]: indent

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/InternalDateExtraFieldTest.java [hunk 20]
```diff
@@ -0,0 +1,330 @@
...
+        @Test
+        void parseFromCentralDirectoryDataShouldReturnLocalDateWhenPassDefaultDateByteArray() throws Exception {
+            InternalDateExtraField testee = new InternalDateExtraField();
+            testee.parseFromCentralDirectoryData(DEFAULT_DATE_BYTE_ARRAY, 0, 8);
+
+            assertThat(testee.getLocalDateValue())
+                .contains(DEFAULT_DATE_LOCAL);
+        }
+    }
+}
```
Comment (rouazana) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/InternalDateExtraFieldTest.java]: I would have expected some specific date test cases in this class

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java (hunk 16 above)
Comment (chibenwa) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: Because you want a fixed time zone (utc) for something stored. What happens if I export my mails from a server in France to import it in a server in Vietnam otherwise.

Hint: you will "loose" 5 hours on all the dates if you only use "getTime".

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java (hunk 17 above)
Comment (chibenwa) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: Then that constructor seems wrong. Maybe we should remove it?

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java (hunk 18 above)
Comment (chibenwa) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: See above explanation please. Your magical "getTime" misses UTC convertion...

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java (hunk 17 above)
Comment (trantienduchn) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: it's used in tests.

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java (hunk 16 above)
Comment (chibenwa) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: Ok, seems like I was wrong on this:

```
Date::getTime
Returns the number of milliseconds since January 1, 1970, 00:00:00 GMT represented by this Date object.
```

The timezone is specified hence relying completely on getTime would work.

Learned something.

Hence I do agree with @rouazana  comments.

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/InternalDateExtraFieldTest.java (hunk 20 above)
Comment (trantienduchn) [reply mailbox/backup/src/test/java/org/apache/james/mailbox/backup/InternalDateExtraFieldTest.java]: I can't get it. You want to have more different days - edge cases as input of our tests?

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java (hunk 16 above)
Comment (rouazana) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: Yes, that's exactly why I meant, thanks.

Comment (trantienduchn) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: But we want to store UTC date time as byte array, it still need conversion from local time to UTC. DOn't you think?

Comment (rouazana) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/InternalDateExtraField.java]: As documented, Date.getTime returns the number of milliseconds since a GMT date. So if the documentation is correct, the conversion is already done calling getTime().
(knowing that UTC ~= GMT)

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/InternalDateExtraFieldTest.java (hunk 20 above)
Comment (rouazana) [reply mailbox/backup/src/test/java/org/apache/james/mailbox/backup/InternalDateExtraFieldTest.java]: In these tests you are just checking that a long is correctly converted to a byte array.
I would like some tests to show that a date is well converted, and that when you get a byte array you get the same date.

Comment (chibenwa) [reply mailbox/backup/src/test/java/org/apache/james/mailbox/backup/InternalDateExtraFieldTest.java]: We miss a test showing that "default date" produces default byte array

**Comment by chibenwa** [comment]
Again the Jenkins bug! `no known node for /home/jenkins/build/workspace/1015`

IMO can be considered :green_apple: ...

**Comment by rouazana** [comment]
merged
